packages/repository/src/relations/has-one/has-one.repository.ts

```diff
@@ -0,0 +1,93 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/example-todo
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Getter} from '@loopback/context';
+import {DataObject, Options} from '../../common-types';
+import {Entity} from '../../model';
+import {Filter, Where} from '../../query';
+import {
+  constrainDataObject,
+  constrainFilter,
+} from '../../repositories/constraint-utils';
+import {EntityCrudRepository} from '../../repositories/repository';
+
+/**
+ * CRUD operations for a target repository of a HasMany relation
+ */
+export interface HasOneRepository<Target extends Entity> {
+  /**
+   * Create a target model instance
+   * @param targetModelData The target model data
+   * @param options Options for the operation
+   * @returns A promise which resolves to the newly created target model instance
+   */
+  create(
+    targetModelData: DataObject<Target>,
+    options?: Options,
+  ): Promise<Target>;
+
+  /**
+   * Find the only target model instance that belongs to the declaring model.
+   * @param filter Query filter without a Where condition
+   * @param options Options for the operations
+   * @returns A promise of the target object or null if not found.
+   */
+  get(
+    filter?: Exclude<Filter<Target>, Where<Target>>,
+    options?: Options,
+  ): Promise<Target | undefined>;
+}
+
+export class DefaultHasOneRepository<
+  TargetEntity extends Entity,
+  TargetID,
+  TargetRepository extends EntityCrudRepository<TargetEntity, TargetID>
+> implements HasOneRepository<TargetEntity> {
+  /**
+   * Constructor of DefaultHasManyEntityCrudRepository
+   * @param getTargetRepository the getter of the related target model repository instance
+   * @param constraint the key value pair representing foreign key name to constrain
+   * the target repository instance
+   */
+  constructor(
+    public getTargetRepository: Getter<TargetRepository>,
+    public constraint: DataObject<TargetEntity>,
+  ) {}
+
+  async create(
+    targetModelData: DataObject<TargetEntity>,
+    options?: Options,
+  ): Promise<TargetEntity> {
+    const targetRepository = await this.getTargetRepository();
+    // should we have an in memory LUT instead of a db query to increase
+    // performance here?
+    const found = await targetRepository.find(
+      constrainFilter({}, this.constraint),
+    );
+    if (found.length > 0) {
+      throw new Error(
+        'HasOne relation does not allow creation of more than one target model instance',
+      );
+    } else {
+      return await targetRepository.create(
```

Comment:
It is crucial to leverage an atomic implementation of `findOrCreate` provided by our connectors. The current proposal is prone to race conditions, where to instances of the target "hasOne" model can be created.

Consider the following scenario: the LB4 server is handling two incoming HTTP requests to create a hasOne target and the code is executed in the following way by Node.js runtime:

1. Request 2 arrives, `DefaultHasOneRepository#create` is invoked
2. `targetRepository.find` is called. It's an async function so other stuff happens while the query is in progress.
3. Request 1  arrives, `DefaultHasOneRepository#create` is invoked.
3. `targetRepository.find` is called. It's an async function so other stuff happens while the query is in progress.
4. `targetRepository.find` for Request 1 returns, no model was found. `targertRepository.create` in called.
5. `targetRepository.find` for Request 2 returns, no model was found. `targertRepository.create` in called.
6.  Both `create` requests eventually finish. We have two models that are a target of `hasOne` relation. 💥 💣 💥  

I am proposing to put this pull request on hold and open a new pull request to add `findOrCreate` method to `EntityCrudRepository` (and the implementations).

Reply:
Thank you for explaining the need for an atomic implementation with a great example. I agree, and I agree with doing that first.

Reply:
I opened a new issue to keep track of "findOrCreate" story, see https://github.com/strongloop/loopback-next/issues/1956